File: sliding_window/best_time_to_buys_and_sell_stock/cpp/best_time_to_buy_and_sell_stock.cpp
#include<iostream>
#include<vector>

using std::vector;

// Time O(n)
// Space O(1)

/* Simple sliding window,
 * always let r - l be optimal by setting l to lowest found
 */
class Solution {
public:
    static int maxProfit(const vector<int>& prices) {
        int l = 0, r = 1;
        int best = 0;

        while(r < prices.size()) {
            prices[r] > prices[l] ?
                best = std::max(best, prices[r] - prices[l]) :
                l = r;
            r++;
        }

        return best;
    }
};

int main() {
    vector<int> prices = { 7, 1, 5, 3, 6, 4 };

    int answer = Solution::maxProfit(prices);

    std::cout << answer << '\n';

    return 0;
}

File: sliding_window/best_time_to_buys_and_sell_stock/js/best_time_to_buy_and_sell_stock.js
const prices = [ 7, 1, 5, 3, 6, 4 ];

const maxProfit = (prices) => {
  let [l, r] = [0, 1];
  let best = 0;

  while(r < prices.length) {
    prices[r] > prices[l] ?
      best = Math.max(best, prices[r] - prices[l]) :
      l = r;
    r++;
  }
  return best
}

console.log(maxProfit(prices));
